File: app/src/main/java/org/stepik/android/adaptive/data/SharedPreferenceMgr.java
package org.stepik.android.adaptive.data;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.gson.Gson;

import org.jetbrains.annotations.NotNull;
import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.stepik.android.adaptive.api.API;
import org.stepik.android.adaptive.api.oauth.OAuthResponse;
import org.stepik.android.adaptive.data.model.AccountCredentials;
import org.stepik.android.adaptive.data.model.Profile;
import org.stepik.android.adaptive.util.Optional;

public final class SharedPreferenceMgr {
    private static final String OAUTH_RESPONSE = "oauth_response";
    private static final String IS_OAUTH_TOKEN_SOCIAL = "is_oauth_token_social";
    private static final String OAUTH_RESPONSE_DEADLINE = "oauth_response_deadline";

    private static final String PROFILE = "profile";
    private static final String PROFILE_ID = "profile_id";

    private static final String NOT_FIRST_TIME = "not_first_time";

    private static final String FAKE_USER = "fake_user";

    private static SharedPreferenceMgr instance;

    private final SharedPreferences sharedPreferences;

    private SharedPreferenceMgr(final Context context) {
        this.sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
    }

    public synchronized static void init(final Context context) {
        if (instance == null) {
            instance = new SharedPreferenceMgr(context);
        }
    }

    public synchronized static SharedPreferenceMgr getInstance() {
        return instance;
    }


    public void saveOAuthResponse(final OAuthResponse response) {
        final Gson gson = new Gson();
        final String json = gson.toJson(response);

        final long currentTime = DateTime.now(DateTimeZone.UTC).getMillis();

        saveString(OAUTH_RESPONSE, json);
        saveLong(OAUTH_RESPONSE_DEADLINE, currentTime + (response.getExpiresIn() - 50) * 1000);
    }

    public OAuthResponse getOAuthResponse() {
        final String json = getString(OAUTH_RESPONSE);
        if (json == null) return null;

        final Gson gson = new Gson();
        return gson.fromJson(json, OAuthResponse.class);
    }

    public void saveProfile(final Profile profile) {
        final Gson gson = new Gson();
        final String json = gson.toJson(profile);

        saveString(PROFILE, json);
        saveLong(PROFILE_ID, profile.getId());
    }

    public Profile getProfile() {
        final String json = getString(PROFILE);
        if (json == null) return null;

        final Gson gson = new Gson();
        return gson.fromJson(json, Profile.class);
    }

    public void removeProfile() {
        API.authLock.lock();
        remove(PROFILE);
        remove(PROFILE_ID);
        remove(OAUTH_RESPONSE);
        remove(IS_OAUTH_TOKEN_SOCIAL);
        remove(OAUTH_RESPONSE_DEADLINE);
        API.authLock.unlock();
    }

    public void removeFakeUser() {
        remove(FAKE_USER);
    }

    public void saveFakeUser(AccountCredentials credentials) {
        final Gson gson = new Gson();
        final String json = gson.toJson(credentials);

        saveString(FAKE_USER, json);
    }

    @NotNull
    public Optional<AccountCredentials> getFakeUser() {
        String json = getString(FAKE_USER);
        if (json == null) return new Optional<>(null);

        return new Optional<>(new Gson().fromJson(json, AccountCredentials.class));
    }

    public void setNotFirstTime(final boolean notFirstTime) {
        saveBoolean(NOT_FIRST_TIME, notFirstTime);
    }

    public boolean isNotFirstTime() {
        return getBoolean(NOT_FIRST_TIME);
    }

    public void setIsOauthTokenSocial(final boolean isOauthTokenSocial) {
        saveBoolean(IS_OAUTH_TOKEN_SOCIAL, isOauthTokenSocial);
    }

    public boolean isAuthTokenSocial() {
        return getBoolean(IS_OAUTH_TOKEN_SOCIAL);
    }

    public long getAuthResponseDeadline() {
        return getLong(OAUTH_RESPONSE_DEADLINE);
    }

    public void resetAuthResponseDeadline() {
        remove(OAUTH_RESPONSE_DEADLINE);
    }

    public long getProfileId() {
        return getLong(PROFILE_ID);
    }

    public void saveBoolean(String name, Boolean data) {
        sharedPreferences.edit().putBoolean(name, data).apply();
    }

    private void saveString(final String name, final String data) {
        sharedPreferences.edit().putString(name, data).apply();
    }

    public void saveLong(final String name, final long data) {
        sharedPreferences.edit().putLong(name, data).apply();
    }

    public long changeLong(final String name, final long delta) {
        final long value = getLong(name) + delta;
        sharedPreferences.edit().putLong(name, value).apply();
        return value;
    }

    private String getString(final String name){
        return sharedPreferences.getString(name, null);
    }

    public long getLong(final String name){
        return sharedPreferences.getLong(name, 0);
    }

    public boolean getBoolean(final String name) {
        return sharedPreferences.getBoolean(name, false);
    }

    public void remove(final String name){
        sharedPreferences.edit().remove(name).apply();
    }

}

File: app/src/main/java/org/stepik/android/adaptive/data/SharedPreferenceMgr.kt
package org.stepik.android.adaptive.data

import android.content.Context
import android.preference.PreferenceManager

import com.google.gson.Gson
import org.joda.time.DateTime
import org.joda.time.DateTimeZone
import org.stepik.android.adaptive.api.Api
import org.stepik.android.adaptive.api.oauth.OAuthResponse
import org.stepik.android.adaptive.content.questions.packs.QuestionsPack
import org.stepik.android.adaptive.data.model.AccountCredentials
import org.stepik.android.adaptive.data.model.Profile
import org.stepik.android.adaptive.di.AppSingleton
import org.stepik.android.adaptive.util.RxOptional
import javax.inject.Inject

@AppSingleton
class SharedPreferenceMgr
@Inject
constructor(context: Context) {

    private val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
    private val gson = Gson()

    var oAuthResponse: OAuthResponse?
        get() {
            val json = getString(OAUTH_RESPONSE) ?: return null
            return gson.fromJson(json, OAuthResponse::class.java)
        }
        set(response) {
            if (response != null) {
                val json = gson.toJson(response)

                val currentTime = DateTime.now(DateTimeZone.UTC).millis

                saveString(OAUTH_RESPONSE, json)
                saveLong(OAUTH_RESPONSE_DEADLINE, currentTime + (response.expiresIn - 50) * 1000)
            }
        }

    var profile: Profile?
        get() {
            val json = getString(PROFILE) ?: return null
            return gson.fromJson(json, Profile::class.java)
        }
        set(profile) {
            if (profile != null) {
                val json = gson.toJson(profile)
                saveString(PROFILE, json)

                profileId = profile.id
            }
        }

    var profileId: Long
        get() = getLong(PROFILE_ID)
        private set(value) {
            saveLong(PROFILE_ID, value)
        }

    val fakeUser: RxOptional<AccountCredentials>
        get() = RxOptional(getString(FAKE_USER)).map { gson.fromJson(it, AccountCredentials::class.java) }

    var isNotFirstTime: Boolean
        get() = getBoolean(NOT_FIRST_TIME)
        set(notFirstTime) = saveBoolean(NOT_FIRST_TIME, notFirstTime)

    val isStreakRestoreTooltipWasShown: Boolean
        get() = getBoolean(IS_STREAK_RESTORE_TOOLTIP_WAS_SHOWN)

    val isPaidContentTooltipWasShown: Boolean
        get() = getBoolean(IS_PAID_CONTENT_TOOLTIP_WAS_SHOWN)

    val isQuestionsPacksTooltipWasShown: Boolean
        get() = getBoolean(IS_QUESTIONS_PACKS_TOOLTIP_WAS_SHOWN)

    var isAuthTokenSocial: Boolean
        get() = getBoolean(IS_OAUTH_TOKEN_SOCIAL)
        set(value) = saveBoolean(IS_OAUTH_TOKEN_SOCIAL, value)

    val authResponseDeadline: Long
        get() = getLong(OAUTH_RESPONSE_DEADLINE)

    var questionsPackIndex: Int
        get() = getInt(QUESTIONS_PACK_INDEX)
        set(value) = saveInt(QUESTIONS_PACK_INDEX, value)

    fun removeProfile() {
        Api.authLock.lock()
        remove(PROFILE)
        remove(PROFILE_ID)
        remove(OAUTH_RESPONSE)
        remove(IS_OAUTH_TOKEN_SOCIAL)
        remove(OAUTH_RESPONSE_DEADLINE)
        Api.authLock.unlock()
    }

    fun removeFakeUser() {
        remove(FAKE_USER)
    }

    fun saveFakeUser(credentials: AccountCredentials) {
        val json = gson.toJson(credentials)
        saveString(FAKE_USER, json)
    }

    fun afterStreakRestoreTooltipWasShown() {
        saveBoolean(IS_STREAK_RESTORE_TOOLTIP_WAS_SHOWN, true)
    }

    fun afterPaidContentTooltipWasShown() {
        saveBoolean(IS_PAID_CONTENT_TOOLTIP_WAS_SHOWN, true)
    }

    fun afterQuestionsPacksTooltipWasShown() {
        saveBoolean(IS_QUESTIONS_PACKS_TOOLTIP_WAS_SHOWN, true)
    }

    fun resetAuthResponseDeadline() {
        remove(OAUTH_RESPONSE_DEADLINE)
    }

    fun onQuestionsPackViewed(pack: QuestionsPack) {
        saveBoolean(QUESTIONS_PACK_VIEWED_PREFIX + pack.id, true)
    }

    fun isQuestionsPackViewed(pack: QuestionsPack): Boolean =
            getBoolean(QUESTIONS_PACK_VIEWED_PREFIX + pack.id)

    fun saveBoolean(name: String, data: Boolean?) {
        sharedPreferences.edit().putBoolean(name, data!!).apply()
    }

    private fun saveString(name: String, data: String) {
        sharedPreferences.edit().putString(name, data).apply()
    }

    fun saveLong(name: String, data: Long) {
        sharedPreferences.edit().putLong(name, data).apply()
    }

    fun changeLong(name: String, delta: Long): Long {
        val value = getLong(name) + delta
        sharedPreferences.edit().putLong(name, value).apply()
        return value
    }

    fun saveInt(name: String, data: Int) {
        sharedPreferences.edit().putInt(name, data).apply()
    }

    private fun getString(name: String): String? = sharedPreferences.getString(name, null)

    fun getLong(name: String): Long = sharedPreferences.getLong(name, 0)

    fun getBoolean(name: String): Boolean = sharedPreferences.getBoolean(name, false)

    fun getInt(name: String): Int = sharedPreferences.getInt(name, 0)

    fun remove(name: String) {
        sharedPreferences.edit().remove(name).apply()
    }

    companion object {
        private const val OAUTH_RESPONSE = "oauth_response"
        private const val IS_OAUTH_TOKEN_SOCIAL = "is_oauth_token_social"
        private const val OAUTH_RESPONSE_DEADLINE = "oauth_response_deadline"

        private const val PROFILE = "profile"
        private const val PROFILE_ID = "profile_id"

        private const val NOT_FIRST_TIME = "not_first_time"

        private const val IS_STREAK_RESTORE_TOOLTIP_WAS_SHOWN = "is_streak_restore_tooltip_was_shown"
        private const val IS_PAID_CONTENT_TOOLTIP_WAS_SHOWN = "is_paid_content_tooltip_was_shown"

        private const val IS_QUESTIONS_PACKS_TOOLTIP_WAS_SHOWN = "is_questions_packs_tooltip_was_shown"

        private const val QUESTIONS_PACK_INDEX = "questions_pack_index"

        private const val FAKE_USER = "fake_user"

        private const val QUESTIONS_PACK_VIEWED_PREFIX = "viewed_"
    }

}